### PR TITLE
IGNITE-13655 Implement readiness probe REST endpoint

### PR DESCRIPTION
**Pull request by akorensh**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/IgnitionEx.java [hunk 1]
```diff
@@ -1476,6 +1476,11 @@ public static DependencyResolver dependencyResolver() {
         return dependencyResolver.get();
     }
 
+    public static boolean hasKernalStarted(String name) {
```
Comment (alamar) [inline modules/core/src/main/java/org/apache/ignite/internal/IgnitionEx.java:1479]: Please add javadoc.

File: modules/core/src/main/java/org/apache/ignite/internal/IgnitionEx.java [hunk 2]
```diff
@@ -3209,6 +3214,13 @@ public void setCounter(int cnt) {
                 this.cnt = cnt;
             }
         }
+
+        /**
+         *
+         * @return whether the startLatch has been counted down, thereby indicating that the kernal has full started.
```
Comment (alamar) [inline modules/core/src/main/java/org/apache/ignite/internal/IgnitionEx.java:3220]: Does it fit in 120 chars?

File: modules/rest-http/pom.xml [hunk 3]
```diff
@@ -146,6 +146,13 @@
             <scope>test</scope>
         </dependency>
 
+        <dependency>
```
Comment (alamar) [inline modules/rest-http/pom.xml:149]: Why is it needed now? Maybe move test to ignite-client module?

File: modules/rest-http/src/test/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/GridProbeCommandTest.java [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+import org.apache.ignite.plugin.AbstractTestPluginProvider;
+import org.apache.ignite.plugin.PluginProvider;
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+import org.junit.Test;
+
+/**
+ * REST probe cmd test.
+ */
+public class GridProbeCommandTest extends GridCommonAbstractTest {
+
```
Comment (alamar) [inline modules/rest-http/src/test/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/GridProbeCommandTest.java:45]: newline not needed, but javadoc is needed (here and below)

File: modules/rest-http/src/test/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/GridProbeCommandTest.java [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+    private static final int JETTY_PORT = 8080;
+
+    private CountDownLatch triggerRestCmdLatch = new CountDownLatch(1);
+
+    private CountDownLatch triggerPluginStartLatch = new CountDownLatch(1);
+
+    public static Map<String, Object> executeProbeRestRequest() throws IOException {
+
```
Comment (alamar) [inline modules/rest-http/src/test/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/GridProbeCommandTest.java:53]: no newline at start of method

File: modules/rest-http/src/test/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/GridRestSuite.java [hunk 6]
```diff
@@ -24,7 +24,8 @@
  */
 @RunWith(Suite.class)
 @Suite.SuiteClasses({
-    RestSetupSimpleTest.class
+    RestSetupSimpleTest.class,
```
Comment (alamar) [inline modules/rest-http/src/test/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/GridRestSuite.java:27]: Do we run this suite anywhere? I guess we don't.

Let's find where the rest of REST tests are (ignite-clients?), move it there.

**Review by alamar — CHANGES_REQUESTED** [review]

File: modules/core/src/main/java/org/apache/ignite/internal/IgnitionEx.java [hunk 7]
```diff
@@ -3209,6 +3214,13 @@ public void setCounter(int cnt) {
...
+         */
+        public boolean hasStartLatchCompleted() { return startLatch.getCount() == 0; }
```
Comment (alamar) [inline modules/core/src/main/java/org/apache/ignite/internal/IgnitionEx.java:3222]: Every please unroll method body to 3 lines. You can import code conventions in your IDEA from https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines

Can you also discuss this approach on the developers list? I'm not sure we should use this latch for checking.

File: modules/core/src/main/java/org/apache/ignite/internal/IgnitionEx.java [hunk 8]
```diff
@@ -1476,6 +1476,11 @@ public static DependencyResolver dependencyResolver() {
         return dependencyResolver.get();
     }
 
+    public static boolean hasKernalStarted(String name) {
+        IgniteNamedInstance grid = name != null ? grids.get(name) : dfltGrid;
+        return grid != null && grid.hasStartLatchCompleted();
```
Comment (alamar) [inline modules/core/src/main/java/org/apache/ignite/internal/IgnitionEx.java:1481]: newline is needed before this line

**Review by alamar — CHANGES_REQUESTED** [review]

File: modules/rest-http/src/test/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/GridRestSuite.java (hunk 6 above)
Comment (akorensh) [reply modules/rest-http/src/test/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/GridRestSuite.java:27]: This is where REST is started. Other REST tests don't actually start http-rest module.

File: modules/core/src/main/java/org/apache/ignite/internal/IgnitionEx.java (hunk 1 above)
Comment (akorensh) [reply modules/core/src/main/java/org/apache/ignite/internal/IgnitionEx.java:1484]: done

File: modules/core/src/main/java/org/apache/ignite/internal/IgnitionEx.java (hunk 2 above)
Comment (akorensh) [reply modules/core/src/main/java/org/apache/ignite/internal/IgnitionEx.java:3224]: yes

**Pull request closed without merging**